### PR TITLE
Handle common floats to string operations in the model and not in the string solver

### PR DESCRIPTION
**Pull request by forejtv**

### CODE REVIEW

File: src/main/java/org/cprover/CProver.java [hunk 1]
```diff
@@ -322,4 +322,17 @@ public static String classIdentifier(Object object)
...
+  /**
+   * This method converts a double to a float and adds assumes that the conversion
+   * did not result any loss of precision. Calling this method is useful when we
+   * need to call a certain complicated operation on double, but we only have it
+   * implemented for floats.
+   */
+  public static float doubleToFloat(double d)
+  {
+    float converted = nondetFloat();
+    CProver.assume(d < Float.MAX_VALUE && d > -Float.MAX_VALUE && d == (double) converted);
```
Comment (romainbrenguier) [inline src/main/java/org/cprover/CProver.java]: This makes sense for test generation but not really for verification. Could we change this assume for an assert?

Comment (romainbrenguier) [inline src/main/java/org/cprover/CProver.java]: Isn't the last condition enough to ensure the two first ones?

Comment (forejtv) [reply src/main/java/org/cprover/CProver.java]: Hmmm. I am not sure what if you suggest is better than the current approach, but I see where you're coming from, at least the user will explicitly see the violated assertion in the output. I've changed this.

**Comment by forejtv** [comment]
Thanks @romainbrenguier for your comments, this is now ready for re-reviews.

File: src/main/java/java/lang/String.java [hunk 2]
```diff
@@ -4266,8 +4266,8 @@ public static String valueOf(float f) {
      * actual program.
      */
     public static String valueOf(double d) {
-        // DIFFBLUE MODEL LIBRARY This is treated internally in CBMC
-        return CProver.nondetWithoutNullForNotModelled();
+        // DIFFBLUE MODEL LIBRARY we cast the number down to float because string solver only knows how to convert floats to string
```
Comment (romainbrenguier) [inline src/main/java/java/lang/String.java]: :pick: this line is long

File: src/main/java/java/lang/String.java [hunk 3]
```diff
@@ -4266,8 +4266,8 @@ public static String valueOf(float f) {
...
+        return valueOf(CProver.doubleToFloat(d));
```
Comment (antlechner) [inline src/main/java/java/lang/String.java:4271]: :question: How is the result of this method call different from just casting `d` to type `float`?

File: src/main/java/java/lang/String.java (hunk 2 above)
Comment (antlechner) [inline src/main/java/java/lang/String.java]: :hammer_and_pick: Could you wrap this line at 80 characters please?
:question: Also, is it a new thing that the string solver only converts floats and not doubles? If not, how did this work before?

File: src/main/java/java/lang/StringBuffer.java [hunk 4]
```diff
@@ -531,11 +531,10 @@ public synchronized StringBuffer append(char c) {
      */
     @Override
     public synchronized StringBuffer append(int i) {
-        // DIFFBLUE MODEL LIBRARY this is replaced internally
         // toStringCache = null;
         // super.append(i);
         // return this;
-        return CProver.nondetWithoutNullForNotModelled();
+        return append(String.valueOf(i));
```
Comment (antlechner) [inline src/main/java/java/lang/StringBuffer.java:537]: :pick: This isn't related to floats so should probably be in a separate commit.
:grey_question: Why is this done for int but not long, char etc.?

**Review by antlechner — COMMENTED** [review]
Could you explain why you're making these changes? Do they improve performance, or was something changed in the string solver that requires this?

File: src/main/java/java/lang/String.java (hunk 3 above)
Comment (forejtv) [reply src/main/java/java/lang/String.java:4271]: In `doubleToFloat` we're ensuring that there's no loss of precision in the conversion, so that we convert the correct number to string. Say we had `double x`, then converted it to `double y`, and then we converted that to `String z`. `z` is a string representation of `y`, but not necessarily of `x`. After a discussion with @romainbrenguier I went for the `doubleToFloat` implementation in order to make it more likely that the string representation of `x` equals the string representation of `y`.

File: src/main/java/java/lang/String.java (hunk 2 above)
Comment (forejtv) [reply src/main/java/java/lang/String.java]: It's not a new thing, before we'd either crash completely or not crash and then return a conversion which was likely to be incorrect.

File: src/main/java/java/lang/StringBuffer.java (hunk 4 above)
Comment (forejtv) [reply src/main/java/java/lang/StringBuffer.java:537]: Good point. It's a follow-up from https://github.com/diffblue/cbmc/pull/4633. I've split this into a separate commit.

Other types should be coming later too, but at the moment I concentrated on the bugs that I've seen in our benchmarks, and am handling them one by one.

File: src/main/java/java/lang/String.java (hunk 3 above)
Comment (antlechner) [reply src/main/java/java/lang/String.java:4271]: I understand the precision problem but I don't understand how `doubleToFloat` avoids this problem. As `float` has fewer bits than `double`, wouldn't there always be a loss of precision? And can we be sure that for every `double d`, there is actually a `float f` such that `d == (double) f;`? Since there are strictly fewer floats than doubles this doesn't look right... and would mean that we sometimes can't get past the `assume` in `doubleToFloat`.
Maybe we could briefly discuss this in person? It's possible that I'm just missing something.
In any case, some tests in another repository would be great to see what this actually does/fixes.

**Comment by forejtv** [comment]
@antlechner The reason for the changes is that there are a lot of `toString` or similar methods in the string solver, and some of them are buggy. To make maintenance easier we decided to use models for most, and along the way fix some of the bug. This is the cbmc PR that removes the string solver hanglind of the methods: https://github.com/diffblue/cbmc/pull/4648

File: src/main/java/org/cprover/CProver.java [hunk 5]
```diff
@@ -322,4 +322,17 @@ public static String classIdentifier(Object object)
...
+  /**
+   * This method converts a double to a float and adds assumes that the conversion
+   * did not result any loss of precision. Calling this method is useful when we
+   * need to call a certain complicated operation on double, but we only have it
+   * implemented for floats.
+   */
+  public static float doubleToFloat(double d)
+  {
+    float converted = nondetFloat();
+    CProver.assume(d == (double) converted);
```
Comment (romainbrenguier) [inline src/main/java/org/cprover/CProver.java]: I would say this makes sense in the models-library but in the core models this assumption shouldn't be there because it constrain the behavior too much and for verification we always want an over-approximation of the behavior.

**Review by antlechner — APPROVED** [review]
Thanks for explaining these changes in person, this makes a lot more sense now I know it's a workaround that will be improved in the future.
It would be great if you could add some documentation about this. Tests in at least one repository would be good too to check if this workaround works for most common values of `double` or if it's very easy to hit the limitations.

**Review by romainbrenguier — APPROVED** [review]
Can you squash the last commit into the relevant one? Otherwise looks good.
